### PR TITLE
PR: Move Online help to new API 

### LINKED ISSUE #12187: Move Online Help plugin to use new API
- [ ] Move to new API
- [ ] Ensure test pass

### PR DESCRIPTION
**Pull request by goanpeca**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Screen Shot 2020-05-26 at 15 01 42

### Issue(s) Resolved

Fixes #12187

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @goanpeca

### CODE REVIEW

**Comment by goanpeca** [comment]
This one is rebased and ready for review @ccordoba12

**Comment by dalthviz** [comment]
Seems like some tests are failing could you check @goanpeca ?

**Comment by goanpeca** [comment]
Will do!

**Comment by goanpeca** [comment]
Fixed @dalthviz

File: spyder/api/widgets/__init__.py [hunk 1]
```diff
@@ -422,7 +425,7 @@ def get_actions(self, filter_actions=True):
             self.dock_action,
             self.undock_action,
             self.close_action,
-            'switch to ' + self._name,
+            self.toggle_view_action,
```
Comment (ccordoba12) [inline spyder/api/widgets/__init__.py:428]: Why is this change needed as part of this PR?

File: spyder/plugins/onlinehelp/plugin.py [hunk 2]
```diff
@@ -7,93 +7,77 @@
...
+    def get_icon(self):
+        return self.create_icon('help')
+
+    def on_close(self, cancelable=False):
         self.save_history()
-        self.set_option('zoom_factor',
-                        self.pydocbrowser.webview.get_zoom_factor())
-        self.pydocbrowser.quit_server()
+
+        # FIXME: move the zoom trigger to an option in the widget?
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/plugin.py:53]: Please address this FIXME

File: spyder/plugins/onlinehelp/plugin.py [hunk 3]
```diff
@@ -7,93 +7,77 @@
...
     def update_font(self):
-        """Reload pydoc browser to get the new font set."""
-        if not PY2:
-            self.pydocbrowser.reload()
+        self.get_widget().reload()
+
+    # --- Public API
+    # ------------------------------------------------------------------------
+    def load_history(self):
+        """Load history from a text file in user home directory."""
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/plugin.py:68]: This docstring is wrong because the file is saved in our config directory.

File: spyder/plugins/onlinehelp/plugin.py [hunk 4]
```diff
@@ -7,93 +7,77 @@
...
+                lines = fh.read().split('\n')
+
+            history = [line.replace('\n', '') for line in lines]
+        else:
+            history = []
+
+        return history
+
+    def save_history(self):
+        """Save history to a text file in user home directory."""
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/plugin.py:80]: This docstring is also wrong.

File: spyder/plugins/onlinehelp/widgets.py [hunk 5]
```diff
@@ -78,123 +101,384 @@ def completer(self):
...
+
+    def _change_url(self, url):
+        """
+        Displayed URL has changed -> updating URL combo box.
+        """
+        self.url_combo.add_text(self.url_to_text(url))
+
+    def _handle_icon_change(self):
+        """
+        Handle icone changes.
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:281]: icone -> icon

File: spyder/plugins/onlinehelp/widgets.py [hunk 6]
```diff
@@ -78,123 +101,384 @@ def completer(self):
...
+            List of searched items.
+        """
+        self.url_combo.addItems(history)
+
+    def initialize(self):
+        """Start pydoc server."""
+        QApplication.setOverrideCursor(QCursor(Qt.WaitCursor))
+        QApplication.processEvents()
+        self.start_server()
+        # Initializing continues in `_initialize_continued` method...
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:313]: This comment doesn't make sense to me. If it's not necessary, please remove it.

File: spyder/plugins/onlinehelp/widgets.py [hunk 7]
```diff
@@ -78,123 +101,384 @@ def completer(self):
...
-    def __init__(self, parent, options_button=None, handle_links=False):
-        WebBrowser.__init__(self, parent, options_button=options_button,
-                            handle_links=handle_links)
+class PydocBrowser(PluginMainWidget):
+    """PyDoc browser widget."""
+
+    DEFAULT_OPTIONS = {
+        'handle_links': False,
+        'max_history_entries': 10,
+        'port': 30128,
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:128]: Is there a need to add this to the widget options? We're not going to offer a UI to modify it, so don't we leave it as constant, like before?

File: spyder/plugins/onlinehelp/widgets.py [hunk 8]
```diff
@@ -78,123 +101,384 @@ def completer(self):
...
+        ----------
+        url: QUrl or str
+            Url address.
+        """
+        self._change_url(url)
+        self.go_to(url)
+
+    def go_to(self, url_or_text):
+        """
+        Go to page *address*.
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:413]: Please change this to

    Go to page URL.

File: spyder/widgets/browser.py [hunk 9]
```diff
@@ -12,24 +12,56 @@
...
+                                    create_plugin_layout, create_toolbutton)
 from spyder.widgets.comboboxes import UrlComboBox
 from spyder.widgets.findreplace import FindReplace
 
 
+# Localization
+_ = get_translation('spyder')
+
+
+# --- Enums
```
Comment (ccordoba12) [inline spyder/widgets/browser.py:38]: Enums -> Constants

File: spyder/widgets/browser.py [hunk 10]
```diff
@@ -52,28 +84,154 @@ def acceptNavigationRequest(self, url, navigation_type, isMainFrame):
...
+        # Actions
+        original_back_action = self.pageAction(QWebEnginePage.Back)
+        back_action = self.create_action(
+            name=WebViewActions.Back,
+            text=_("Back"),
+            icon=self.create_icon('previous'),
+            triggered=lambda: original_back_action.trigger(),
+            context=Qt.WidgetWithChildrenShortcut,
+        )
+        original_forward_action = self.pageAction(QWebEnginePage.Forward)
```
Comment (ccordoba12) [inline spyder/widgets/browser.py:117]: Please add a blank line above this one to follow the same convention you established in the lines below.

File: spyder/widgets/browser.py [hunk 11]
```diff
@@ -52,28 +84,154 @@ def acceptNavigationRequest(self, url, navigation_type, isMainFrame):
...
+
+        original_refresh_action = self.pageAction(QWebEnginePage.Reload)
+        self.create_action(
+            name=WebViewActions.Refresh,
+            text=_("Refresh"),
+            icon=self.create_icon('refresh'),
+            triggered=lambda: original_refresh_action.trigger(),
+            context=Qt.WidgetWithChildrenShortcut,
+        )
+        original_stop_action = self.pageAction(QWebEnginePage.Stop)
```
Comment (ccordoba12) [inline spyder/widgets/browser.py:177]: Blank line above this one.

File: spyder/widgets/browser.py [hunk 12]
```diff
@@ -52,28 +84,154 @@ def acceptNavigationRequest(self, url, navigation_type, isMainFrame):
...
+        self.add_item_to_menu(
+            self.zoom_out_action,
+            menu=menu,
+            section=WebViewMenuSections.Zoom,
+        )
+        self.add_item_to_menu(
+            inspect_action,
+            menu=menu,
+            section=WebViewMenuSections.Extras,
+        )
```
Comment (ccordoba12) [inline spyder/widgets/browser.py:222]: Please use a `for` to simplify this code.

File: spyder/plugins/onlinehelp/widgets.py (hunk 7 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:128]: Because it is an option of the widget. Having it exposed as a configuration is not really relevant.

File: spyder/api/widgets/__init__.py (hunk 1 above)
Comment (goanpeca) [reply spyder/api/widgets/__init__.py:428]: Nothing specific to this PR, but it was something that needed fixing in the API

File: spyder/plugins/onlinehelp/plugin.py (hunk 2 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/plugin.py:53]: Deleting comment. Not fixable since changing that will not emit a signal.

File: spyder/plugins/onlinehelp/plugin.py (hunk 3 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/plugin.py:68]: Fixed

File: spyder/plugins/onlinehelp/plugin.py (hunk 4 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/plugin.py:80]: Fixed

File: spyder/plugins/onlinehelp/widgets.py (hunk 7 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:128]: To further elaborate, since this dict is used to store options/settings from the widgte, the option of leaving a constant is not appealing because:

`class PydocBrowser(PluginMainWidget): ``

1.) Having an attribute in the class means that now we have 2 different ways of handling options, some might be on the DEFAULTS_OPTIONS and some are class attributes. This is confusing and could enforce a bad practice.

2.) Having a constant outside the class might be better since now it is not a class attribute, so we are not incurring in 1, but now it lives outside the widget.

So I suggest we do not do that to enforce cleaner code, even if the option is not likely to change

File: spyder/plugins/onlinehelp/widgets.py (hunk 5 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:281]: Fixed

File: spyder/plugins/onlinehelp/widgets.py (hunk 6 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:313]: 🤷

File: spyder/plugins/onlinehelp/widgets.py (hunk 8 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:413]: Fixed

File: spyder/widgets/browser.py (hunk 9 above)
Comment (goanpeca) [reply spyder/widgets/browser.py:38]: Fixed

File: spyder/widgets/browser.py (hunk 10 above)
Comment (goanpeca) [reply spyder/widgets/browser.py:117]: Fixed

File: spyder/widgets/browser.py (hunk 11 above)
Comment (goanpeca) [reply spyder/widgets/browser.py:177]: Fixed

File: spyder/widgets/browser.py (hunk 12 above)
Comment (goanpeca) [reply spyder/widgets/browser.py:222]: Fixed (but it needs several for cycles, since the items are being added to different sections)

**Comment by goanpeca** [comment]
Added the requested fixes plus answers to your questions @ccordoba12

File: spyder/plugins/onlinehelp/widgets.py (hunk 7 above)
Comment (ccordoba12) [reply spyder/plugins/onlinehelp/widgets.py:128]: > some might be on the DEFAULTS_OPTIONS and some are class attributes

This is really not a class attribute. It's a constant, so why not simply declare it as such at the beginning of the file?

**Comment by ccordoba12** [comment]
I left one comment about the default port used by pydoc (which I think it's a constant and not a class attribute) and tests are still failing.

**Comment by goanpeca** [comment]
>I left one comment about the default port used by pydoc (which I think it's a constant and not a class attribute) and 

Fixed

>tests are still failing.

The failing tests spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_get_pydoc[lib0] FAILED [ 67%] is passing on my mac machine. Not sure what might be the problem.

Screen Shot 2020-06-11 at 21 00 55

**Comment by goanpeca** [comment]
Ok, I removed the skip that was there on master. These tests were not passing on Unix https://github.com/spyder-ide/spyder/blob/master/spyder/plugins/onlinehelp/tests/test_pydocgui.py#L40

Now they pass on windows and linux

File: spyder/plugins/onlinehelp/plugin.py [hunk 13]
```diff
@@ -10,90 +10,85 @@
...
+                             self.get_widget().get_zoom_factor())
+        return True
 
-        layout = QVBoxLayout()
-        layout.addWidget(self.pydocbrowser)
-        self.setLayout(layout)
+    def register(self):
+        widget = self.get_widget()
+        widget.load_history(self.load_history())
+        widget.sig_load_finished.connect(self.sig_load_finished)
```
Comment (dalthviz) [inline spyder/plugins/onlinehelp/plugin.py:67]: Maybe changing this line for something like `widget.sig_load_finished.connect(self.sig_load_finished.emit)` could fix the failing test @goanpeca ?

Comment (goanpeca) [reply spyder/plugins/onlinehelp/plugin.py:67]: >Maybe changing this line for something like widget.sig_load_finished.connect(self.sig_load_finished.emit) could fix the failing test @goanpeca ?

It passes locally on my mac, and those tests were failing before this PR and were skipped on Linux and Mac. Also signals should not normally need to have that in place, they coual automatically emit if connected as pointed out. But with what you suggest I will give a try to

`widget.sig_load_finished.connect(lambda: self.sig_load_finished())`  however, it seems more likely that something is blocking the server from starting, maybe the port :shrug: ?

Anyway, now test pass on Linux and win, which is better than the case before :-)

File: spyder/plugins/onlinehelp/widgets.py [hunk 14]
```diff
@@ -4,25 +4,50 @@
...
+
+
+# Localization
+_ = get_translation('spyder')
+
+
+# --- Constants
+# ----------------------------------------------------------------------------
+PORT = 30128
+PORT = 7464
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:40]: There are two port definitions here. Please leave only the first one.

File: spyder/plugins/onlinehelp/widgets.py [hunk 15]
```diff
@@ -46,155 +71,421 @@ def spyder_safeimport(path, forceload=0, cache={}):
...
+        """
         string_url = url.toString()
         if 'about:blank' in string_url:
             return 'about:blank'
         elif 'get?key=' in string_url or 'search?key=' in string_url:
-            return ''
-        return osp.splitext(to_text_string(url.path()))[0][1:]
+            return url.toString().split('=')[-1]
+
+        return os.path.splitext(str(url.path()))[0][1:]
```
Comment (ccordoba12) [inline spyder/plugins/onlinehelp/widgets.py:393]: Please use `osp` instead of `os.path` here.

File: spyder/plugins/onlinehelp/widgets.py (hunk 14 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:40]: Oops, will do!

Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:40]: Fixed

File: spyder/plugins/onlinehelp/widgets.py (hunk 15 above)
Comment (goanpeca) [reply spyder/plugins/onlinehelp/widgets.py:393]: Fixed
